### DISCUSSION TITLE
What is the reason to UNIT TEST like use test framework such Jest or Vitest, what is the benefit ?

### DISCUSSION THREAD

**Discussion by gaomingzhao666** [community/community#128236] [Programming Help] [answered]
### Select Topic Area

Question

### Hey guys, how's going? I just wondering about why we need to use a test framework and UNIT TEST for apps.

**I watched the documentation of the framework like Jest, and I'm confused about why we don't use console.log to displace UNIT TEST ?**

**We can just test a function by printing a value and judge that type of it but why do we need to write another method like `describe()` and `expect()`  and even ugly need to import it**

**Compare the code in the following, I can not see any profit to using a test framework, Is there anybody tell me why ? it's so weird !!!**

- Jest example

``` js
const sum = require('./sum');

test('adds 1 + 2 to equal 3', () => {
  expect(sum(1, 2)).toBe(3);
});
```

- log example

```js
console.log(sum(1, 2)); // 3
```

**Comment by ebndev** [comment]
Hi @gaomingzhao666, thanks for posting in GitHub Discussions! 

I've gone ahead and moved this post to our Programming Help category, as this topic might be more relevant there and hopefully someone give you a nudge in the right direction 😄

**Reply by gaomingzhao666** [reply]
thank you

**Comment by anko** [comment, marked as answer]
I agree that a framework is often overkill. What matters is that the tests are useful. If you can do that without a library, great.

One benefit of testing frameworks however is that they give helpful errors when tests fail. A test framework for the test you listed would typically generates output like:

    In test 'adds 1 + 2 to equal 3': Unexpected value `4`. (Expected `3`.)

Some even use a highlighted diff, to help spot differences in large compared objects.

In contrast:

- If you just use logging, you would have to check the results manually.

- If you use a function call like [`console.assert`](https://developer.mozilla.org/en-US/docs/Web/API/console/assert_static) that simply errors if the check fails, you will get a less specific error:

   ```js
   console.assert(1 + 2 == 4)
   // → Assertion failed: 
   ```

  Which might be fine in simple cases.
  
  But if you wanted output similar to Jest's, you may have to type a lot more:

   ```js
   const result = sum(1, 2);
   const expected = 4;
   console.assert(result == expected, { result, expected });
   //→ Assertion failed: Object { result: 3, expected: 4 }
   ```

   You could extract that into its own function, like `assert_equals`, to reuse it! And soon you'll find you've implemented a testing library. :sweat_smile:

**Reply by gaomingzhao666** [reply]
**so appreciate your help !!**

**But if I use a test framework and test a UNIT, the code seems too much, is that valuable to do it? OR I just use a test framework in a size-not-match project, how scale of a project could make use of a test framework a good deal ?**

**Reply by gaomingzhao666** [reply]
**Just because too much code for test purposes probably can cause bad readability, so I am not sure that is valuable if I use a test framework. 😵‍💫**
